### DISCUSSION TITLE
$refs magic array empty in teleported element

### DISCUSSION THREAD

**Discussion by sergio0983** [alpinejs/alpine#3933] [7. Bugs (archived)]
Hi. 

I found that attempting to access this.$refs in a teleported element  won´t work.

A sample code that highlights the issue: 

```
<div
    class="relative"
    x-data="{ 
	isOpen: false,
	toggle:(){
		if (this.isOpen) {
                    this.close();
                }
                else {
                    this.open();
                }
	},
        open() {
            this.isOpen = true;
            this.$refs.trigger.setAttribute('aria-expanded', 'true');
            this.$refs.panel.focus();
        },
        close() {
            this.isOpen = false;
            this.$refs.trigger.setAttribute('aria-expanded', 'false');
        },
    }"
>
	<div
        @click="toggle()"
        class="flex items-center cursor-pointer"
        x-ref="trigger"
    >
        {{ $trigger }}
    </div>
	
	<template x-teleport="{{ $teleportDestination }}">
        <div
            x-ref="panel"
            x-cloak 
            x-show="isOpen"
            @click.away="close()"
            class="absolute z-10 w-48 py-2 mt-2 bg-white rounded-md shadow-xl"
			x-transition
        >
            {{ $content }}
        </div>
    </template>
</div>
```
Since close() is called from within the teleported element, this.$refs.trigger is undefined, hence, causing an error. 

As a side note, if I use this.$root._x_refs instead of this.$refs it works. 

By looking at the code at **packages/alpinejs/src/magics/$refs.js**, I think that a possible solution would be changing the getArrayOfRefObject method; if it is a teleported node, instead of searching for x_refs on itself and its parent, search for them on itself, the root object, and the parents of the root object.

**Comment by edvordo** [comment]
Can confirm this is an issue, reproducible with something as simple as
```html
<div x-data="{ 
  toggleItem() { 
    console.log({ r: this.$refs.test }) // { r: undefined }
   }
}">
    <div>
        <input name="test" value="test" x-ref="test" />
    </div>
    <template x-teleport="body">
        <button type="button" class="absolute top-1/2 left-1/2" x-on:click="toggleItem()">click me</button>
    </template>
</div>
```
The docs, as of writing this comment state
> Teleported content can access the normal Alpine scope of the component as well as other features like $refs, $root, etc...

But `$refs` are indeed empty. 

This worked in v 2.8.1.

The workaround by using `$this.$root._x_refs` suggested by @sergio0983 works, so thank you for that.

**Reply by ekwoka** [reply]
What if you omit the () from the expression?

**Reply by edvordo** [reply]
> What if you omit the () from the expression?

The result is the same, `this.$refs` is still empty.
